### PR TITLE
Add Managed Version Info on hover #238

### PR DESCRIPTION
**Pull request by vrubezhny**
Signed-off-by: Victor Rubezhny <vrubezhny@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
Just a simple comment when you create a PR. I think it should be nice to add a little screenshot with th eresult of PR (or a little demo). It will be easier to understand what the PR provides and for history it is nice too.

We do that for LemMinX and I find it is nice for the reviewer.

It is just a suggestion that I do.

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	public void testManagedVersionHoverInPluginParent() throws IOException, URISyntaxException {
+		DOMDocument document = createDOMDocument("/pom-pluginManagement-parent.xml", languageService);
+		Hover hover = languageService.doHover(document, new Position(14, 33), new SharedSettings());
+		String value = ((MarkupContent) hover.getContents().getRight()).getValue();
+		System.out.println("testManagedVersionHoverInPluginParent: [" + value + "]");
+		assertNotNull(value);
+		assertTrue(value.contains("The managed version is"));
+		assertTrue(value.contains("2.22.2"));
+		assertTrue(value.contains("The artifact is managed in"));
+		assertTrue(value.contains("org.apache.maven.plugins:maven-surefire-plugin:2.22.2"));
```
Comment (mickaelistria) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java:101]: I don't think this should be added in when the version is the one that is defined in the pom. Managed version are only useful when no version is explicitly set (usually in the "child pom")

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	public void testManagedVersionHoverInPluginChild() throws IOException, URISyntaxException {
+		DOMDocument document = createDOMDocument("/pom-pluginManagement-child.xml", languageService);
+		Hover hover = languageService.doHover(document, new Position(18, 33), new SharedSettings());
+		String value = ((MarkupContent) hover.getContents().getRight()).getValue();
+		System.out.println("testManagedVersionHoverInPluginChild: [" + value + "]");
+		assertNotNull(value);
+		assertTrue(value.contains("The managed version is"));
+		assertTrue(value.contains("2.22.2"));
+		assertTrue(value.contains("The artifact is managed in"));
+		assertTrue(value.contains("org.apache.maven.plugins:maven-surefire-plugin:2.22.2"));
```
Comment (mickaelistria) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java:87]: I think what is expected here is more the reference to the parent pom that adds the version management (here pluginManagement-parent.xml)

**Review by mickaelistria — CHANGES_REQUESTED** [review]
Unless I missed something, I believe some extra informations are confusing and should either be fixed or removed.

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java (hunk 1 above)
Comment (vrubezhny) [reply lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java:101]: But we still have to show the version and the location links on a pom where the version is clearly defined.

That was actually a problem as the method I've used for finding version in such case was returning a wrong version and link. So, I'd prefer this test case to also exist for parent pom.

Comment (mickaelistria) [reply lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java:101]: > But we still have to show the version and the location links on a pom where the version is clearly defined.

we already have that I believe. Ctrl+Click can be used to get the link to the artifact.

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	public void testManagedVersionHoverInDependencyChiild() throws IOException, URISyntaxException {
+		DOMDocument document = createDOMDocument("/pom-dependencyManagement-child.xml", languageService);
+		Hover hover = languageService.doHover(document, new Position(15, 25), new SharedSettings());
+		String value = ((MarkupContent) hover.getContents().getRight()).getValue();
+		System.out.println("testManagedVersionHoverInDependencyChiild: [" + value + "]");
+		assertNotNull(value);
+		assertTrue(value.contains("The managed version is"));
+		assertTrue(value.contains("2.22.2"));
+		assertTrue(value.contains("The artifact is managed in"));
+		assertTrue(value.contains("org.apache.maven.plugins:maven-surefire-plugin:2.22.2"));
```
Comment (mickaelistria) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java:60]: That's not correct. The artifact version is managed in pom-dependencyManagement-parent.

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	@Test
+	@Timeout(90000)
+	public void testManagedVersionHoverInDependencyParent() throws IOException, URISyntaxException {
+		DOMDocument document = createDOMDocument("/pom-dependencyManagement-parent.xml", languageService);
+		Hover hover = languageService.doHover(document, new Position(13, 29), new SharedSettings());
+		String value = ((MarkupContent) hover.getContents().getRight()).getValue();
+		System.out.println("testManagedVersionHoverInDependencyParent: [" + value + "]");
+		assertNotNull(value);
+		assertFalse(value.contains("The managed version is"));
+		assertFalse(value.contains("The artifact is managed in"));
```
Comment (mickaelistria) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java:72]: good.

**Review by mickaelistria — CHANGES_REQUESTED** [review]

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java (hunk 3 above)
Comment (vrubezhny) [reply lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java]: Now a version from the parent pom and location of parent pom where the version is defined is shown in Hover.

If no version is specified in project it  shows a version and location of pom for an according artifact resolved in maven project.

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java (hunk 2 above)
Comment (vrubezhny) [reply lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/ManagedVersionHoverTest.java]: Fixed

**Comment by mickaelistria** [comment]
Thanks!

**Comment by pbodnar** [comment]
Thank you!

It would be great to also jump directly to the line where the given artifact version is defined and not to just open the linked pom.xml. A new feature request targeted on LemMinX, on m2e, or maybe both?

**Pull request merged**